var/spack/repos/builtin/packages/foam-extend/package.py

```diff
@@ -40,207 +130,234 @@ class FoamExtend(Package):
+        'WM_COMPILER':  None,
+        'WM_OPTIONS':   None,
+    }
+
+    build_script = './spack-Allwmake'  # <- Generated by patch() method.
+    # phases = ['configure', 'build', 'install']
 
-    depends_on('paraview', when='+paraview')
+    def setup_environment(self, spack_env, run_env):
+        run_env.set('FOAM_INST_DIR', self.prefix)
+        run_env.set('WM_PROJECT_DIR', self.projectdir)
+
+    @property
+    def _canonical(self):
+        """Canonical name for this package and version"""
+        return 'foam-extend-{0}'.format(self.version.up_to(2))
+
+    @property
+    def projectdir(self):
+        """Absolute location of project directory: WM_PROJECT_DIR/"""
+        return join_path(self.prefix, self._canonical)  # <- prefix/canonical
+
+    @property
+    def etc(self):
+        """Absolute location of the OpenFOAM etc/ directory"""
+        return join_path(self.projectdir, 'etc')
+
+    @property
+    def archbin(self):
+        """Relative location of architecture-specific executables"""
+        wm_options = self.set_openfoam()
+        return join_path('applications', 'bin', wm_options)
+
+    @property
+    def archlib(self):
+        """Relative location of architecture-specific libraries"""
+        wm_options = self.set_openfoam()
+        return join_path('lib', wm_options)
+
+    @property
+    def mplib_content(self):
+        """Define 'SPACK' mpi settings to have wmake
+        use spack information with minimum modifications to OpenFOAM
+        """
+        return [
+            'PFLAGS = -DOMPI_SKIP_MPICXX -DMPICH_IGNORE_CXX_SEEK',
+            'PINC   = -I{0}'.format(self.spec['mpi'].prefix.include),
+            'PLIBS  = -L{0} -lmpi'.format(pkglib(self.spec['mpi']))
+        ]
+
+    @property
+    def rpath_info(self):
+        """Define 'SPACKOpt' compiler optimization file to have wmake
+        use spack information with minimum modifications to OpenFOAM
+        """
+        build_libpath   = join_path(self.stage.source_path, self.archlib)
+        install_libpath = join_path(self.projectdir, self.archlib)
+
+        # 'DBUG': rpaths
+        return '{0}{1} {2}{3}'.format(
+            self.compiler.cxx_rpath_arg, install_libpath,
+            self.compiler.cxx_rpath_arg, build_libpath,
+        )
+
+    # For foam-extend: supports darwinIntel, does not yet support +int64
+    def set_openfoam(self):
+        """Populate foam_cfg, foam_sys according to
+        variants, architecture, compiler.
+        Returns WM_OPTIONS.
+        """
+        # Run once
+        wm_options = self.foam_sys['WM_OPTIONS']
+        if wm_options:
+            return wm_options
+
+        # Trivial check
+        if self.compiler.name not in self.supported_compilers:
+            raise RuntimeError
+            (
+                '{0} is an unsupported compiler'.format(self.compiler.name)
+            )
 
-    def set_arch(self):
         (sysname, nodename, release, version, machine) = os.uname()
```

Comment:
This probably isn't going to work on cross-compiled platforms.  You should get the target information for the build from the spec you're passed in `install()`.  So, if you want Spack's machine info, you should do this:

```python
# This is a general machine type.  Spack supports the notion
# of platforms with multiple OSs and targets (like Cray and BG/Q).
# e.g.: linux, darwin, cray, bgq
platform = spec.architecture.platform

# e.g., x86_64, haswell, ppc64le
target = spec.architecture.target

# could be better named.  e.g., redhat6, redhat7, ubuntu14, yosemite
os = spec.architecture.platform_os
```

Getting this information from `uname` is going to get you info about the login node.  On Cray and BG/Q machines, that is not where you're going to run the code -- it's just where you'll build it.

Here's an example.  On a Cray we can say:

```console
spack install foo os=backend target=haswell
```

This builds for the default backend OS, e.g. compute node linux (CNL) on Cray.  `haswell` is the default target on machines like Cori, but there are also targets like `mic-knl` if you want to compile for Knights Landing.  You'll be handed a spec in install with an architecture triple like `cray-cnl-haswell`.

```console
spack install foo os=frontend 
```
This builds for whatever the default frontend (login node) OS is, with a default target.  Usually that is something like `sles12` and `x86_64`.  

Anyway -- TL; DR, you should get your machine info from the spec.  For Linux and Darwin, you can assume for the moment that your target is pretty much what uname would give you, though we do plan to support more specialized targets like `haswell` on regular Linux boxes, too -- see #3206.